File: 0399-evaluate-division/0399-evaluate-division.cpp
class Solution {
public:
    double bfs(string start, string end, unordered_map<string, vector<pair<string, double>>> adjList) {
        unordered_set<string> vis;
        queue<pair<string, double>> q;
        q.push({start, 1});
        vis.insert(start);
        while (!q.empty()) {
            string node1 = q.front().first;
            double wt1 = q.front().second;
            q.pop();
            if (node1 == end) return wt1;
            for (auto nei: adjList[node1]) {
                if (!vis.count(nei.first)) {
                    vis.insert(nei.first);
                    double wt2 = nei.second;
                    q.push({nei.first, wt1 * wt2});
                }
            }
        }
        return -1;
    }

    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        int n = equations.size();
        unordered_map<string, vector<pair<string, double>>> adjList;
        int i = 0;
        for (auto e: equations) {
            adjList[e[0]].push_back({e[1], values[i]});
            adjList[e[1]].push_back({e[0], (((double)1) / values[i])});
            i++;
        }
        vector<double> res;
        for (auto q: queries) {
            if (!adjList.count(q[0]) || !adjList.count(q[1])) {
                res.push_back(-1);
                continue;
            }
            if (q[1] == q[0]) {
                res.push_back(1);
                continue;
            }
            double ans = bfs(q[0], q[1], adjList);
            res.push_back(ans);
        }
        return res;
    }
    
};

File: 0399-evaluate-division/0399-evaluate-division.cs
public class Solution {
    private Dictionary<string,List<(string c,double w)>> adj;

    public double[] CalcEquation(IList<IList<string>> equations, double[] values, IList<IList<string>> queries) {
        adj = new Dictionary<string,List<(string c,double w)>>();
        int i = 0;
        foreach (var eq in equations) {
            double wt = values[i++];
            string src = eq[0];
            string dst = eq[1];

            if (!adj.ContainsKey(src)) adj[src] = new List<(string c, double w)>();
            if (!adj.ContainsKey(dst)) adj[dst] = new List<(string c, double w)>();

            adj[src].Add((dst, wt));
            adj[dst].Add((src, 1 / wt));
        }
        var res = new double[queries.Count];

        i = 0;
        foreach (var q in queries) {
            if (!adj.ContainsKey(q[0]) || !adj.ContainsKey(q[1])) res[i++] = -1;
            else if (q[0] == q[1]) res[i++] = 1;
            else res[i++] = bfs(q[0], q[1]);
        }
        return res;
    }

    public double bfs(string src, string dst) {
        var q = new Queue<(string c, double w)>();
        var visited = new HashSet<string>();
        visited.Add(src);
        q.Enqueue((src, 1));
        while (q.Count != 0) {
            var node = q.Dequeue();
            if (node.c == dst) return node.w;
            foreach (var nei in adj[node.c]) {
                if (!visited.Contains(nei.c)) {
                    visited.Add(nei.c);
                    q.Enqueue((nei.c, node.w * nei.w));
                }
            }
        }
        return -1;
    }









}